lib/Hakyll/Core/Configuration.hs

```diff
@@ -132,3 +142,7 @@ shouldIgnoreFile conf path = orM
     orM :: [IO Bool] -> IO Bool
     orM []       = return False
     orM (x : xs) = x >>= \b -> if b then return True else orM xs
+
+shouldWatchIgnore :: Configuration -> FilePath -> FilePath -> IO Bool
+shouldWatchIgnore conf path providerDir = (|| watchIgnore conf path') <$> shouldIgnoreFile conf path
+  where path' = makeRelative providerDir path
```

Comment:
`Configuration` already contains a path to the provider directory, so the third parameter here makes the function  error-prone: if the user passes something that'd different from the config, the result won't match Hakyll's internal notion of what should be ignored, and that will confuse the user.

We could export two versions of this function: this one, and one that only takes 2 parameters and runs `canonicalizePath` internally. That'd work, but it's weird to export things that are only meant for internal use.

I did a quick benchmark of `canonicalizePath` and on my Linux box it takes about 35 mcs for a 100-chars-long filename that doesn't exist. At 30,000 files, it'd take 1 second just to check the names, which isn't too good. So we can't just drop the third parameter and run `canonicalizePath` on every invocation.

The best solution I can think of is `Configuration -> IO (FilePath -> IO Bool)`. It'd run `canonicalizePath` once, and return a closure which can quickly check as many files as needed.

What do you think? Do you see any alternative solutions?

(Also, this function should have a docstring ;)

Reply:
@Minoru 
> `Configuration` already contains a path to the provider directory, so the third parameter here makes the function error-prone

You're very right, thanks for double checking this!

> The best solution I can think of is Configuration -> IO (FilePath -> IO Bool). It'd run canonicalizePath once, and return a closure which can quickly check as many files as needed.

I like this solution a lot, I implemented a version of it. I'm not sure what the naming scheme is for functions returning functions, so I just left it as `shouldWatchIgnore`, please check if that fits.

I added the docstring too, somehow it escaped my mind that this function will be visible to users.

Thanks for the quick review!